### PR TITLE
TF Updates

### PR DESCRIPTION
**Pull request by farizrahman4u**

### CODE REVIEW

**Comment by farizrahman4u** [comment]
@AlexDBlack I cant seem to reproduce dep issue, rest of the issues should be gone now.

File: nd4j/nd4j-tensorflow/src/main/java/org/nd4j/tensorflow/conversion/graphrunner/GraphRunner.java [hunk 1]
```diff
@@ -435,27 +308,30 @@ public GraphRunner(SavedModelConfig savedModelConfig, ConfigProto sessionOptions
...
-                    //outputs
+                    //outputSchema
                     outputOut, outputTensorsPointer, savedModelConfig.getSaveModelOutputOrder().size(),
                     //targets
                     null, 0,
                     null,
-                    status);
+                    status);        long end = System.nanoTime();
+            long diff = TimeUnit.NANOSECONDS.toMillis((end - start));
+            log.info("Session run  timing in ms " + diff);
```
Comment (AlexDBlack) [inline nd4j/nd4j-tensorflow/src/main/java/org/nd4j/tensorflow/conversion/graphrunner/GraphRunner.java]: Did we want to log this at info level?
I could imagine a busy web server having a lot of log spam...
`log.debug("Session run  timing in ms {}", diff);`
Same thing in another method below.

**Review by AlexDBlack — DISMISSED** [review]
LGTM, other than maybe minor tweak to logging level in a couple of places

**Comment by AlexDBlack** [comment]
> @AlexDBlack I cant seem to reproduce dep issue, rest of the issues should be gone now.

Still seeing the windows gson dependency issue, other issues confirmed fixed. I'll look into that.

**Comment by AlexDBlack** [comment]
Confirmed passing now on windows + linux for CPU

CUDA is failing, however with `java.lang.RuntimeException: cudaMemcpyAsync failed` on most of the tests.
To reproduce: run tests in nd4j-tests-tensorflow with nd4j-tf-gpu and tf-gpu profiles.


For example:
```
java.lang.RuntimeException: cudaMemcpyAsync failed

	at org.nd4j.jita.allocator.pointers.cuda.cudaStream_t.synchronize(cudaStream_t.java:41)
	at org.nd4j.jita.flow.impl.SynchronousFlowController.commitTransfer(SynchronousFlowController.java:315)
	at org.nd4j.jita.handler.impl.CudaZeroHandler.relocate(CudaZeroHandler.java:406)
	at org.nd4j.jita.handler.impl.CudaZeroHandler.getDevicePointer(CudaZeroHandler.java:727)
	at org.nd4j.jita.allocator.impl.AtomicAllocator.getPointer(AtomicAllocator.java:325)
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.naiveExec(CudaExecutioner.java:384)
	at org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner.exec(CudaExecutioner.java:572)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.equalsWithEps(BaseNDArray.java:4374)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.equals(BaseNDArray.java:4430)
	at org.nd4j.linalg.jcublas.JCublasNDArray.equals(JCublasNDArray.java:507)
	at org.junit.Assert.isEquals(Assert.java:131)
	at org.junit.Assert.equalsRegardingNull(Assert.java:127)
	at org.junit.Assert.assertEquals(Assert.java:111)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.nd4j.tensorflow.conversion.TensorflowConversionTest.testConversionFromNdArray(TensorflowConversionTest.java:59)
```



CUDA tests are failing for me on _master_ though with the following:
```
java.lang.UnsatisfiedLinkError: no jnitensorflow in java.library.path

	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1543)
...
Caused by: java.lang.UnsatisfiedLinkError: C:\Users\Alex\.javacpp\cache\tensorflow-1.15.0-1.5.2-windows-x86_64-gpu.jar\org\bytedeco\tensorflow\windows-x86_64-gpu\jnitensorflow.dll: Can't find dependent libraries
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1493)
```

**Comment by AlexDBlack** [comment]
OK, so I'm seeing most tests pass now on CUDA (windows)

If I run GraphRunnerTest.testGraphRunnerFilePath in isolation, or run it as part of all tests in GraphRunnerTest, I get the following failure:
```
java.lang.AssertionError: 
Expected :[    2.0000,    4.0000,    6.0000,    8.0000]
Actual   :[         0,         0,         0,         0]
```

However, if I right click on `org.nd4j.tensorflow.conversion` under `gpujava` and run tests, I get this:

Note all tests are running, and GraphRunnerTest.testGraphRunnerFilePath passes - but GpuGraphRunnerTest.testGraphRunner now fails with exactly the same issue.

GpuGraphRunnerTest.testGraphRunner also fails when run in isolation.

image


Note that a `Nd4j.getExecutioner().commit();` and/or `Thread.sleep(2000)` after `graphRunner.run(inputs)` doesn't make any difference.

**Comment by agibsonccc** [comment]
@farizrahman4u

**Comment by AlexDBlack** [comment]
@farizrahman4u Any updates/progress here?

**Comment by farizrahman4u** [comment]
@AlexDBlack on it

**Comment by farizrahman4u** [comment]
:shipit: yes

**Comment by farizrahman4u** [comment]
@AlexDBlack try now?

**Review by AlexDBlack — APPROVED** [review]
I'm not following _how_ this was fixed (we previously talked about adding opyToHost ops), but I can confirm that tests are consistently passing for me now on CUDA.

**Comment by farizrahman4u** [comment]
I got curious and created a buffer from the device pointer. As I suspected, it had the same behavior as the host array - the first test case of a test suit always fails and the rest passes, even if there is a test identical to the first one. (Which also mean test cases would fail when run individually, because each one is the "first" in that session). So I simply added a "warm up" functionality to emulate this.

**Review by AlexDBlack — APPROVED** [review]

**Pull request merged**